crates/hc_sandbox/src/save.rs

```diff
@@ -79,18 +79,20 @@ pub fn clean(mut hc_dir: PathBuf, sandboxes: Vec<usize>) -> anyhow::Result<()> {
 
 /// Load sandbox paths from the `.hc` file.
 pub fn load(mut hc_dir: PathBuf) -> anyhow::Result<Vec<PathBuf>> {
-    let mut paths = Vec::new();
     hc_dir.push(".hc");
-    if hc_dir.exists() {
-        let existing = std::fs::read_to_string(hc_dir)?;
-        for sandbox in existing.lines() {
-            let path = PathBuf::from(sandbox);
-            let config_file_path = ConfigFilePath::from(ConfigRootPath::from(path.clone()));
-            if config_file_path.as_ref().exists() {
-                paths.push(path);
-            } else {
-                tracing::error!("Failed to load path {} from existing .hc", path.display());
-            }
+    if !hc_dir.is_file() {
+        return Ok(Vec::new());
+    };
+    let mut paths = Vec::new();
+    let existing = std::fs::read_to_string(&hc_dir)
+        .with_context(|| format!("Failed to read file: {}", hc_dir.display()))?;
+    for sandbox in existing.lines() {
+        let path = PathBuf::from(sandbox);
+        let config_file_path = ConfigFilePath::from(ConfigRootPath::from(path.clone()));
+        if config_file_path.as_ref().exists() {
```

Comment:
Part of the problem has already happened here. If any paths are filtered out then the indices that the user might be using to start sandboxes have been invalidated.

For example, if I `hc s generate -n 3 my.happ` and then delete the directory for the sandbox at index 1 then this code filters out that directory when I do `hc s run <n>` and I will be able to do `hc s run 0` just fun but `hc s run 1` will run the sandbox that is actually at index 2 in the `.hc` file. Trying to do `hc s run 2` won't refer to a valid index at all.

So any checks about the user's input after this point are less useful in my mind. Even if they provided an index that appears valid, it might not be what they meant.